Deal with DMI busy in block reads/writes

timsifive opened the pull request:
This was trivial for writes, but more painful for reads. In doing so (I think) I simplified the block read code a little, although it's still a bit messy.

I finally went through the effort to test this properly. To do that I added a `riscv reset_delays` command. Combined with https://github.com/riscv/riscv-isa-sim/pull/265 and some riscv-tests changes which I'll push once these 2 have merged, I now get coverage of all the corner cases in the memory progbuf functions by running `make all` in riscv-tests/debug.

timsifive committed:
Deal with DMI busy in block reads/writes.

Rename dtmcontrol_idle to dtmcs_idle, because the register is now called
dtmcs.
Simplify read_memory_progbuf_inner(), getting rid of several unnecessary
variables.

Change-Id: Ibac655a45c63cf2210ab282568b54f8097526c10

timsifive committed:
Add `riscv reset_delays` for testing.

This allows me to test corner cases in block read/write errors.

Change-Id: I3ccfe707851dbc578277ea0d5e278eab81a3c7ef

timsifive committed:
Fix block read corner cases.

Change-Id: I841f264ca881078075beaa58023dd0e0a81f3ff3

mwachs5 commented on src/target/riscv/riscv-013.c:
```diff
@@ -503,14 +513,20 @@ static dmi_status_t dmi_scan(struct target *target, uint32_t *address_in,
-static int dmi_op_timeout(struct target *target, uint32_t *data_in, int dmi_op,
-		uint32_t address, uint32_t data_out, int timeout_sec)
+/* If dmi_busy_encountered is non-NULL, this function will use it to tell the
+ * caller whether DMI was every busy during this call. */
```
every -> ever

timsifive replied:
Fixed.

timsifive committed:
Fix typo.

Change-Id: Ibdd26c5c524b10a3518fe708e9b7fc917b0cb1b6

timsifive commented:
Can I get a review/approval for this?

SiFiveGregS commented:
I'll take a look today.  It reaches deeper RISCV debug architecture-level details than I've reviewed before, which is good (for me).

GregSavin commented on src/target/riscv/riscv-011.c:
```diff
@@ -206,6 +206,8 @@ typedef struct {
 	bool need_strict_step;
 	bool never_halted;
+
+	int reset_delays_wait;
```
I wonder if a code comment generally describing the purpose of reset_delays_wait might be worthwhile here, and in riscv013_info_t.  Although I do see the command help text for the reset_delays Tcl command, which explains the same thing at a different level.

timsifive replied:
I'll add comments.

GregSavin commented on src/target/riscv/riscv-013.c:
```diff
@@ -2095,31 +2104,43 @@ static int read_memory_progbuf_inner(struct target *target, target_addr_t addres
-	dmi_write(target, DMI_ABSTRACTAUTO,
-			1 << DMI_ABSTRACTAUTO_AUTOEXECDATA_OFFSET);
+	if (count == 1) {
+		uint64_t value;
```
Is this "if" block a new optimization special case?

timsifive replied:
Yes, but it's also essential after the rewrite because I didn't add all the code below that's necessary to properly work when there's just 1 word being read.

timsifive committed:
Add comment for reset_delays_wait.

Also refactor so there's just one of them in riscv, instead of one for
0.11 and one for 0.13.

Change-Id: I0dbbf112b4c57f76bed971a22dadf844fa27cd4e
